Controller test times out waiting for socket readiness

## Summary

`TestControllerShutdown` is failing in local verification with:

```text
controller_test.go:165: timed out waiting for controller socket to become available
```

I hit this while verifying unrelated migration work; I did not touch controller code. The failure reproduces consistently on branch `feat/migrate-doctor` and appears to be an existing controller startup/readiness problem rather than a regression from the migration changes.

I also saw a likely-related command-level symptom in `cmd/gc/testdata/controller.txtar`: `gc stop` returned `City stopped.` but the script expected `Controller stopping`, suggesting either output drift or a readiness/race around controller lifecycle behavior.

## Repro

From repo root:

```bash
go test ./cmd/gc -run TestControllerShutdown -count=1 -timeout 120s -v
```

Observed:

```text
=== RUN   TestControllerShutdown
    controller_test.go:165: timed out waiting for controller socket to become available
--- FAIL: TestControllerShutdown (5.01s)
FAIL
FAIL	github.com/gastownhall/gascity/cmd/gc	5.828s
FAIL
```

## Relevant Code

The failing wait is here:

- `cmd/gc/controller_test.go:165` calls `waitForController(t, dir, 5*time.Second)`

That helper polls `controllerAlive(...)` until the Unix socket responds to `ping`.

The controller socket is started very early in `runController(...)`:

- `cmd/gc/controller.go` calls `startControllerSocket(...)` before token generation, controller state setup, API startup, and the main runtime loop.

So if this timeout is real, one of these is likely happening:

1. `runController` exits before or during socket startup and the test never sees why
2. socket listen succeeds but the server goroutine never becomes responsive to `ping`
3. startup is taking longer than expected under some conditions and the 5s readiness timeout is too aggressive
4. there is a race/cleanup issue around the socket file or lock in temp dirs

## Why I think this is not caused by my branch

- The branch changes were in migration/doctor/main command registration, not controller runtime
- I saw the same failure when running the isolated controller test directly
- Earlier, a full `go test ./... -json -count=1 -timeout 600s` also failed only on this controller test

## Related Symptom

In one earlier broader test run, `cmd/gc/testdata/controller.txtar` also failed:

```text
> exec gc stop $WORK/ctrl-city
[stdout]
City stopped.
> stdout 'Controller stopping'
FAIL: testdata/controller.txtar:13: no match for `Controller stopping` found in stdout
```

That may be a separate output expectation drift, but it might also help correlate whether others are seeing the same controller lifecycle/readiness problem.

## Questions / next checks

- Is this a known flake in `TestControllerShutdown`?
- Should `waitForController` be more diagnostic and fail with controller stdout/stderr if startup never happens?
- Should `runController` surface early-startup errors to the test more directly?
- Is the command-level output in `controller.txtar` stale relative to current `gc stop` behavior?

--- a/cmd/gc/cmd_stop_test.go
+++ b/cmd/gc/cmd_stop_test.go
@@ -52,7 +52,8 @@ func TestCmdStopWaitsForStandaloneControllerExit(t *testing.T) {
 	}
 	const seededSession = "seeded-session"
 
-	var controllerStdout, controllerStderr bytes.Buffer
+	var controllerStdout bytes.Buffer
+	var controllerStderr syncBuffer
 	done := make(chan struct{})
 	go func() {
 		runController(dir, tomlPath, cfg, "", buildFn, nil, sp, nil, nil, nil, nil, events.Discard, nil, &controllerStdout, &controllerStderr)
@@ -70,7 +71,7 @@ func TestCmdStopWaitsForStandaloneControllerExit(t *testing.T) {
 		}
 	})
 
-	waitForController(t, dir, 5*time.Second)
+	waitForController(t, dir, 5*time.Second, done, &controllerStderr)
 	if err := sp.Start(context.Background(), seededSession, runtime.Config{}); err != nil {
 		t.Fatal(err)
 	}

--- a/cmd/gc/controller_test.go
+++ b/cmd/gc/controller_test.go
@@ -7,6 +7,7 @@ import (
 	"os"
 	"path/filepath"
 	"strings"
+	"sync"
 	"sync/atomic"
 	"testing"
 	"time"
@@ -140,7 +141,8 @@ func TestControllerShutdown(t *testing.T) {
 	// Dolt-backed .beads/ database).
 	tomlPath := writeCityTOML(t, dir, "test", "mayor")
 
-	var stdout, stderr bytes.Buffer
+	var stdout bytes.Buffer
+	var stderr syncBuffer
 
 	// Run controller in a goroutine; it will block until canceled.
 	// Use a close-able channel so cleanup can detect whether the
@@ -162,7 +164,7 @@ func TestControllerShutdown(t *testing.T) {
 	})
 
 	// Poll for controller socket to become available instead of fixed sleep.
-	waitForController(t, dir, 5*time.Second)
+	waitForController(t, dir, 5*time.Second, done, &stderr)
 
 	if !tryStopController(dir, &bytes.Buffer{}) {
 		t.Fatal("tryStopController returned false, expected true")
@@ -814,7 +816,8 @@ func TestControllerPokeTriggersImmediate(t *testing.T) {
 	// operations rather than falling back to cwd.
 	tomlPath := writeCityTOML(t, dir, "test")
 
-	var stdout, stderr bytes.Buffer
+	var stdout bytes.Buffer
+	var stderr syncBuffer
 
 	done := make(chan struct{})
 	go func() {
@@ -832,7 +835,7 @@ func TestControllerPokeTriggersImmediate(t *testing.T) {
 	})
 
 	// Poll for controller socket to become available.
-	waitForController(t, dir, 5*time.Second)
+	waitForController(t, dir, 5*time.Second, done, &stderr)
 
 	// Wait for initial tick.
 	deadline := time.After(5 * time.Second)
@@ -875,19 +878,66 @@ func TestControllerPokeTriggersImmediate(t *testing.T) {
 	}
 }
 
+// syncBuffer is a concurrency-safe bytes.Buffer for use as an io.Writer
+// (e.g. capturing stderr from a goroutine) that can be read safely from
+// another goroutine. It implements io.Writer plus String/Len accessors.
+type syncBuffer struct {
+	mu  sync.Mutex
+	buf bytes.Buffer
+}
+
+func (sb *syncBuffer) Write(p []byte) (int, error) {
+	sb.mu.Lock()
+	defer sb.mu.Unlock()
+	return sb.buf.Write(p)
+}
+
+func (sb *syncBuffer) String() string {
+	sb.mu.Lock()
+	defer sb.mu.Unlock()
+	return sb.buf.String()
+}
+
+func (sb *syncBuffer) Len() int {
+	sb.mu.Lock()
+	defer sb.mu.Unlock()
+	return sb.buf.Len()
+}
+
 // waitForController polls until the controller socket at dir is responsive,
-// or fails the test after the given timeout. This replaces fixed sleeps that
-// are unreliable under load.
-func waitForController(t *testing.T, dir string, timeout time.Duration) {
+// or fails the test after the given timeout. If done is non-nil it is checked
+// on each poll iteration; a closed channel means runController exited early
+// and the real error is in stderr rather than a socket timeout.
+func waitForController(t *testing.T, dir string, timeout time.Duration, done <-chan struct{}, stderr *syncBuffer) {
 	t.Helper()
 	deadline := time.After(timeout)
 	for {
 		if controllerAlive(dir) != 0 {
 			return
 		}
+		// Detect early exit: runController returned before the socket
+		// became responsive. Report stderr so the real error surfaces
+		// instead of a misleading "timed out" message.
+		if done != nil {
+			select {
+			case <-done:
+				var diag string
+				if stderr != nil {
+					diag = stderr.String()
+				}
+				t.Fatalf("controller exited before socket became ready; stderr: %s", diag)
+			default:
+			}
+		}
 		select {
 		case <-deadline:
-			t.Fatal("timed out waiting for controller socket to become available")
+			msg := "timed out waiting for controller socket to become available"
+			if stderr != nil {
+				if s := stderr.String(); s != "" {
+					msg += "; stderr: " + s
+				}
+			}
+			t.Fatal(msg)
 		default:
 			time.Sleep(10 * time.Millisecond)
 		}

--- a/cmd/gc/controller_test.go
+++ b/cmd/gc/controller_test.go
@@ -3,8 +3,10 @@ package main
 import (
 	"bytes"
 	"context"
+	"fmt"
 	"net"
 	"os"
+	"os/exec"
 	"path/filepath"
 	"strings"
 	"sync"
@@ -880,7 +882,7 @@ func TestControllerPokeTriggersImmediate(t *testing.T) {
 
 // syncBuffer is a concurrency-safe bytes.Buffer for use as an io.Writer
 // (e.g. capturing stderr from a goroutine) that can be read safely from
-// another goroutine. It implements io.Writer plus String/Len accessors.
+// another goroutine. It implements io.Writer plus a String accessor.
 type syncBuffer struct {
 	mu  sync.Mutex
 	buf bytes.Buffer
@@ -896,12 +898,6 @@ func (sb *syncBuffer) String() string {
 	sb.mu.Lock()
 	defer sb.mu.Unlock()
 	return sb.buf.String()
-}
-
-func (sb *syncBuffer) Len() int {
-	sb.mu.Lock()
-	defer sb.mu.Unlock()
-	return sb.buf.Len()
 }
 
 // waitForController polls until the controller socket at dir is responsive,
@@ -941,6 +937,69 @@ func waitForController(t *testing.T, dir string, timeout time.Duration, done <-c
 		default:
 			time.Sleep(10 * time.Millisecond)
 		}
+	}
+}
+
+// TestWaitForControllerDetectsEarlyExit verifies that waitForController
+// surfaces the real stderr content (not a generic timeout) when the
+// controller goroutine exits before the socket becomes ready.
+func TestWaitForControllerDetectsEarlyExit(t *testing.T) {
+	if os.Getenv("TEST_CONTROLLER_EARLY_EXIT") == "1" {
+		dir := t.TempDir()
+
+		var stderr syncBuffer
+		fmt.Fprint(&stderr, "gc start: injected startup failure\n") //nolint:errcheck // test setup
+
+		done := make(chan struct{})
+		close(done) // controller already exited
+
+		waitForController(t, dir, 5*time.Second, done, &stderr)
+		return
+	}
+
+	cmd := exec.Command(os.Args[0], "-test.run=^TestWaitForControllerDetectsEarlyExit$")
+	cmd.Env = append(os.Environ(), "TEST_CONTROLLER_EARLY_EXIT=1")
+	out, err := cmd.CombinedOutput()
+	if err == nil {
+		t.Fatal("expected subprocess to fail via t.Fatalf")
+	}
+	output := string(out)
+	if !strings.Contains(output, "controller exited before socket became ready") {
+		t.Fatalf("expected early-exit diagnostic, got:\n%s", output)
+	}
+	if !strings.Contains(output, "injected startup failure") {
+		t.Fatalf("expected stderr content in diagnostic, got:\n%s", output)
+	}
+}
+
+// TestWaitForControllerTimeoutIncludesStderr verifies that when
+// waitForController times out, it appends any buffered stderr content
+// to the failure message for diagnosis.
+func TestWaitForControllerTimeoutIncludesStderr(t *testing.T) {
+	if os.Getenv("TEST_CONTROLLER_TIMEOUT_STDERR") == "1" {
+		dir := t.TempDir()
+
+		var stderr syncBuffer
+		fmt.Fprint(&stderr, "gc start: partial startup output\n") //nolint:errcheck // test setup
+
+		done := make(chan struct{}) // never closed — simulates hung controller
+
+		waitForController(t, dir, 50*time.Millisecond, done, &stderr)
+		return
+	}
+
+	cmd := exec.Command(os.Args[0], "-test.run=^TestWaitForControllerTimeoutIncludesStderr$")
+	cmd.Env = append(os.Environ(), "TEST_CONTROLLER_TIMEOUT_STDERR=1")
+	out, err := cmd.CombinedOutput()
+	if err == nil {
+		t.Fatal("expected subprocess to fail via t.Fatalf")
+	}
+	output := string(out)
+	if !strings.Contains(output, "timed out waiting for controller socket") {
+		t.Fatalf("expected timeout message, got:\n%s", output)
+	}
+	if !strings.Contains(output, "partial startup output") {
+		t.Fatalf("expected stderr appended to timeout message, got:\n%s", output)
 	}
 }
 

